volatility3/framework/symbols/linux/__init__.py

```diff
@@ -419,3 +430,342 @@ def get_module_from_volobj_type(
         kernel = context.modules[kernel_module_name]
 
         return kernel
+
+    @classmethod
+    def choose_kernel_tree(cls, vmlinux: interfaces.context.ModuleInterface) -> "Tree":
+        """Returns the appropriate tree data structure instance for the current kernel implementation.
+        This is used by the IDR and the PageCache to choose between the XArray and RadixTree.
+
+        Args:
+            vmlinux: The kernel module object
+
+        Returns:
+            The appropriate Tree instance for the current kernel
+        """
+        address_space_type = vmlinux.get_type("address_space")
+        address_space_has_i_pages = address_space_type.has_member("i_pages")
+        i_pages_type_name = (
+            address_space_type.child_template("i_pages").vol.type_name
+            if address_space_has_i_pages
+            else ""
+        )
+        i_pages_is_xarray = i_pages_type_name.endswith(constants.BANG + "xarray")
+        i_pages_is_radix_tree_root = i_pages_type_name.endswith(
+            constants.BANG + "radix_tree_root"
+        ) and vmlinux.get_type("radix_tree_root").has_member("xa_head")
+
+        if i_pages_is_xarray or i_pages_is_radix_tree_root:
+            return XArray(vmlinux)
+        else:
+            return RadixTree(vmlinux)
+
+
+class Tree(ABC):
```

Comment:
The name's a little generic.  The module helps identify it, but if there's a slightly more descriptive name than Tree, it would be good to just differentiate it a little.  There are other operating systems in here too, and we don't want all of them to have their own tree (plus, I'm quite sure linux will have different tree structures too).  5;P

Reply:
Yeah, sorry I know. Actually, 'Tree' isn't quite right since XArray is not a tree. I had a hard time coming up with a good name. Basically, it's an abstract base class which allows selecting and executing the Linux kernel RadixTree and XArray data structures. Earlier kernel versions used RadixTree in various subsystems, but many of these have been replaced with XArrays.  Since these serve various functions in the kernel, it seems best to keep the name generic either way. What do you think of something like DSAbstract or DSBase (DS => Data Structure)?, any better?

Reply:
Ok, renamed to IDStorage as it represents what all of them store, which is identifiers.